### PR TITLE
Removed `.isel` for `DatasetRolling.construct` consistent rolling behavior

### LINKED ISSUE #7021: Inconsistent behavior between `DatasetRolling.construct` and `DataArrayRolling.construct` with stride > 1.
### What is your issue?

INSTALLED VERSIONS
------------------
commit: None
python: 3.8.10 | packaged by conda-forge | (default, May 11 2021, 07:01:05) 
[GCC 9.3.0]
python-bits: 64
OS: Linux
OS-release: 5.4.0-73-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: 4.9.0

xarray: 2022.6.0
pandas: 1.4.2
numpy: 1.19.5
scipy: 1.7.0
netCDF4: 1.6.0
pydap: None
h5netcdf: 1.0.2
h5py: 3.1.0
Nio: None
zarr: 2.12.0
cftime: 1.6.1
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: 1.3.4
dask: 2021.06.2
distributed: 2021.06.2
matplotlib: 3.5.3
cartopy: None
seaborn: 0.12.0
numbagg: None
fsspec: 2021.07.0
cupy: 9.2.0
pint: None
sparse: 0.13.0
flox: None
numpy_groupies: None
setuptools: 49.6.0.post20210108
pip: 21.1.3
conda: 4.10.3
pytest: 6.2.4
IPython: 7.24.1
sphinx: None
## Reproducing the problem
I have an xarray Dataset with a single dimension as specified. (Or any arbitrary  Xarray's `Dataset`
```
> Dimensions:
> time: 11058688
```
When applied rolling operation on `DataArray` with no overlapping window, it is working as one would normally expected.
```python
dataset.var_a.rolling(k=256).construct('w', stride=256)
```
`11058688 / 256 = 43198`
```
> Dimensions:
> time: 43198, k:  256   # 43198 windows
```

However when applying the same operation to the `Dataset`:

```python
dataset.rolling(k=256).construct('w', stride=256)
```
```
> Dimensions:
> time: 169, k:  256   # How can we even arrived at 169 windows?
```

I don't see any reasons why should rolling on `Dataset` and `DataArray` should behave differently. Shouldn't rolling on dataset is just repeating `DataArray` rolling on every data variable?
This differing behavior is not mentioned on the documentation either.

### PR DESCRIPTION
**Pull request by p4perf4ce**
`Dataset(...).isel(...)` at the return caused `DatasetRolling.construct` behavior to be inconsistent with `DataArrayRolling.construct` when `stride` > 1 without any benefits.

The bug was reported in #7021

- [x] Closes #7021
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

**Comment by headtr1ck** [comment]
Good spot. The fix seems good to me.
If you fix the tests and add a whats-new this is good to go

**Comment by headtr1ck** [comment]
Sorry for the late reply...
This PR is still not providing the correct solution.

For some reason, when you add a dimension-coordinate to the Dataset it behaves as expected (before) and now wrong (with this PR):

E.g. use these two test datasets:
```python
ds1 = xr.Dataset({"x": ("t", range(20))})
ds2 = xr.Dataset({"x": ("t", range(20))}, {"t": range(20)})

print("Dataset rolling: ds1")
print(ds1.rolling(t=4).construct("w", stride=2).x.values)  # wrong in main
print("DataArray rolling: ds1")
print(ds1.x.rolling(t=4).construct("w", stride=2).values)
print("Dataset rolling: ds2")
print(ds2.rolling(t=4).construct("w", stride=2).x.values)  # wrong in this PR
print("DataArray rolling: ds2")
print(ds2.x.rolling(t=4).construct("w", stride=2).values)
```

The unit tests use datasets with an dimension-coordinate, therefore this error was never spotted.

**Comment by headtr1ck** [comment]
After digging a bit more, the problematic line is this one:
`Dataset(dataset, coords=self.obj.coords ...`
Here, dataset is a `dict` of constructed DataArrays (they have the correct shape) which have the strided "t" index and `self.obj.coords` are the original coordinates which contain the unstrided "t" index.
Upon Dataset creation, these are getting broadcasted and NaNs are inserted into the constructed DataArrays.

The isel in the end is supposed to remove the inserted NaNs again.

So I think we have to find an intermediate solution and remove the isel + adopt what we pass to coords.

**Comment by headtr1ck** [comment]
Nvmd. I have added another test with more dimensions and 2D coordinates.
My approach does not work here :/
> cannot reindex or align along dimension 't' because of conflicting dimension sizes: {10, 20} (note: an index is found along that dimension with size=10)

Anyone knows how to align here properly? Coords do not have an isel, otherwise one could simply apply the stride as well.

**Comment by headtr1ck** [comment]
For now the approach is to stride the original dataset and then extract the coords from there. Ofc, this strides the dataset variables which are then not used, so unnecessary computation. However this approach is already much faster and memory efficient than the previous approach.

**Comment by p4perf4ce** [comment]
My apologies for very late reply. Got tons of backlog until seeing this popped up in my mailbox.

> The isel in the end is supposed to remove the inserted NaNs again.

Thank you for digging this out. I was dumbfounded when looking at this particular line, haven't thought about NaNs case back then.

> For now the approach is to stride the original dataset and then extract the coords from there. Ofc, this strides the dataset variables which are then not used, so unnecessary computation. However this approach is already much faster and memory efficient than the previous approach.

I've ended up with something similar but a little bit different in my own internal repository. I've found that it's a bit more efficient and more practical to just create a class of **_Virtual_**  rolling coordinate then accessing the data by asking the virtual coordinate to provide me a `coord` at window `i`. So float64 dataset with 10++ dimensions (I'm dealing with biosignal datasets) won't explode my memory (only coords that grew), but it is a little bit too radical of a change for the current approach that both I proposed originally, and you have done so far. 

But it would turn your question of how to not stride over `coord` again to just doing things only on `coord` and left the rest as is.

**Comment by headtr1ck** [comment]
Not sure I understand what you mean.

The current approach only temporarily strides the dataset including it's coords and then extracts those coords. 
Since it does not align the result to the non-strided indexes anymore it should be much more memory efficient.

Unfortunately the Coordinates class does not support indexing, so we have to do it at the dataset level. I think it should not add too much overhead because it is index based lookup.

The main difference between this approach and what you did is that it supports coordinates that have different dimensions than the data variables (see the new test).

**Comment by headtr1ck** [comment]
Actually we could add a peakmem asv benchmark for this and see how much more.memory efficient it is.

**Comment by headtr1ck** [comment]
> I think this is within the expectation? Because original behavior causes the result to fall short by a large margin (see my issue at https://github.com/pydata/xarray/issues/7021). Now that this PR fixing it, number of result windows should be larger (thus larger memory footprint) when running benchmark against the mainline branch.

Your behavior was for Datasets without an dimension coordinate (a coordinate that is called the same as the dimension), the benchmark uses a Dataset with (otherwise I cannot compare the results correctly).

This behavior was correct before, because what happened is that the created arrays were first strided, then extended to full again filling the missing data with NaNs and then strided again (this was the isel in the end).

So I expect the extending to full part to consume more memory, but seeing in the benchmark it does apparently not.

Anyway I think this PR is a good addition because it fixes a bug, which is far more important than performance.


> Thank you for your helpful feedback!

You're welcome!

File: xarray/core/rolling.py [hunk 1]
```diff
@@ -785,11 +785,14 @@ def construct(
             if not keep_attrs:
                 dataset[key].attrs = {}
 
+        # Need to stride coords as well. TODO: is there a better way?
+        coords = self.obj.isel(
```
Comment (dcherian) [inline xarray/core/rolling.py:789]:
```suggestion
        coords = self.obj.coords.to_dataset().isel(
```

Perhaps this is slightly better?

Comment (p4perf4ce) [reply xarray/core/rolling.py:789]: I've tried running asv locally based on your suggestion. I've try increased `nx` and `nt` to 3000 and 10000 respectively to see if there's a gap.  I think I don't see much difference here.
New:
```
[33.33%] ··· rolling.DatasetRollingMemory.peakmem_1drolling_construct                                                                                                                                                                                      ok
[33.33%] ··· ======== ======
              stride
             -------- ------
               None    577M
                5      583M
                50     578M
             ======== ======

[66.67%] ··· rolling.DatasetRollingMemory.peakmem_1drolling_reduce                                                                                                                                                                                         ok
[66.67%] ··· ====== ====== =======
             --     use_bottleneck
             ------ --------------
              func   True   False
             ====== ====== =======
              sum    581M    1.6G
              max    575M   1.61G
              mean   575M    1.6G
             ====== ====== =======

[100.00%] ··· rolling.DatasetRollingMemory.peakmem_ndrolling_reduce                                                                                                                                                                                         ok
[100.00%] ··· ====== ====== =======
              --     use_bottleneck
              ------ --------------
               func   True   False
              ====== ====== =======
               sum    595M   1.61G
               max    596M   1.62G
               mean   618M   1.61G
              ====== ====== =======
```

Old:
```
[33.33%] ··· rolling.DatasetRollingMemory.peakmem_1drolling_construct                                                                                                                                                                                      ok
[33.33%] ··· ======== ======
              stride
             -------- ------
               None    577M
                5      577M
                50     578M
             ======== ======

[66.67%] ··· rolling.DatasetRollingMemory.peakmem_1drolling_reduce                                                                                                                                                                                         ok
[66.67%] ··· ====== ====== =======
             --     use_bottleneck
             ------ --------------
              func   True   False
             ====== ====== =======
              sum    576M    1.6G
              max    575M    1.6G
              mean   575M    1.6G
             ====== ====== =======

[100.00%] ··· rolling.DatasetRollingMemory.peakmem_ndrolling_reduce                                                                                                                                                                                         ok
[100.00%] ··· ====== ====== =======
              --     use_bottleneck
              ------ --------------
               func   True   False
              ====== ====== =======
               sum    604M   1.62G
               max    601M   1.61G
               mean   618M   1.62G
              ====== ====== =======
```

Comment (headtr1ck) [reply xarray/core/rolling.py:789]: > Perhaps this is slightly better?

Then I get:
> ValueError: Dimensions {'index'} do not exist. Expected one or more of Frozen({})

**Comment by headtr1ck** [comment]
I just increased the dimensions as well in CI and now we get:

```
| Change   | Before [b08a9d86]    | After [7359559b]    |   Ratio | Benchmark (Parameter)                                        |
+ grep 'Traceback \|failed\|PERFORMANCE DECREASED' benchmarks.log
|----------|----------------------|---------------------|---------|--------------------------------------------------------------|
| -        | 5.57G                | 204M                |    0.04 | rolling.DatasetRollingMemory.peakmem_1drolling_construct(5)  |
| -        | 5.59G                | 204M                |    0.04 | rolling.DatasetRollingMemory.peakmem_1drolling_construct(50) |
```
So it got much better? This seems strange...
But anyway, I think this is good to go.

**Comment by p4perf4ce** [comment]
> So it got much better? This seems strange...

Just to note. It still almost impossible to run construct with stride>1 on a large dataset even with the first commit in this PR (only fix my wanted behavior) because it ended up not creating a view but actually allocating memory (according to memray) for rolling windows, so I thought that this was an intended behavior (leading to my reply suggesting some virtual lookup things I used elsewhere). My assumption is something weird is going on when extending to full part since reduce doesn't suffer the same issue. I might help digging down on this later but doesn't seem to be an immediate issue.

Anyway, somehow it ends up fixing both behavior bug and performance. Thank you everyone.

**Comment by headtr1ck** [comment]
Thanks for starting this :) Your insight was helpful in figuring out what was going wrong.

Just a tip: Do future PRs on a branch in your forked repo and not the main branch. Since we do squash commits, your history will be divergent to xarrays main branch and you will have to force push. Doing this in a branch prevents this :)

Lets wait a day or two and then merge :)

**Review by mathause — APPROVED** [review]
LGTM!

**Pull request merged**